sklearn/preprocessing/tests/test_data.py

```diff
@@ -1535,9 +1528,17 @@ def _check_one_hot(X, X2, cat, n_features):
     assert_array_equal(toarray(B), toarray(D))
 
 
+def test_one_hot_encoder_string():
+    X = [['cat', 'domestic'], ['wolf', 'wild']]
+    enc = OneHotEncoder()
+    enc.fit(X)
+    Xtr = enc.transform([['cat', 'wild']])
+    assert_array_equal(toarray(Xtr), [[1, 0, 0, 1]])
+
+
 def test_one_hot_encoder_categorical_features():
     X = np.array([[3, 2, 1], [0, 1, 1]])
-    X2 = np.array([[1, 1, 1]])
+    X2 = np.array([[3, 1, 1]])
```

Comment:
Consider the following code
```python
from sklearn.preprocessing import OneHotEncoder
import numpy as np

enc = OneHotEncoder(handle_unknown='error')
X = np.array([[3], [5], [7]])
enc.fit(X)
print(enc.transform([[4]]).toarray())
```

### Output on master
```
[[ 0.  0.  0.]]
```

### Output on this branch
```
Traceback (most recent call last):
  File "test.py", line 9, in <module>
    print(enc.transform([[4]]).toarray())
  File "/home/vighnesh/git/scikit-learn/sklearn/preprocessing/data.py", line 1845, in transform
    selected=self.categorical_features)
  File "/home/vighnesh/git/scikit-learn/sklearn/preprocessing/data.py", line 1657, in _apply_selected
    return transform(X)
  File "/home/vighnesh/git/scikit-learn/sklearn/preprocessing/data.py", line 1882, in _transform
    raise ValueError(msg)
ValueError: Unknown feature(s) [4] in column 0
```

The current implementation in master any value of a feature less than or equal to its maximum value during fit is acceptable when `n_values="auto"`. With the above code run on the master branch, any value between `[0,7]` will not throw an error, in spite of the fact that it was never seen during fit.

To support strings I have to make the assumption that any value of a feature supplied during transform should either have been seen during fit, or specified explicitly using `classes/values` argument. This will break existing code like the one I have written above.